content/blog/2023-04-04-the-pci-device-that-wasnt.md

```diff
@@ -0,0 +1,289 @@
+
+Deep in the guts of the device subsystem, if we have a look at `device_add()` we find the following:
+
+```c
+int device_add(struct device *dev)
+{
+        /* lots of device init stuff... */
+
+        if (dev->bus)
+                blocking_notifier_call_chain(&dev->bus->p->bus_notifier,
+                                             BUS_NOTIFY_ADD_DEVICE, dev);
+```
+
+If the device we're initialising is attached to a bus, then we call the bus notifier of that bus with the `BUS_NOTIFY_ADD_DEVICE` notification, and the device in question.  So we're going through the process of adding a VIO device, and somehow calling into a notifier that's only registered for PCI devices.  I did a bunch of debugging to see if our VIO device was somehow malformed and pointing to a PCI bus, or the `struct subsys_private` (that's the `bus->p` above) was somehow pointing to the wrong place, but everything seemed sane.  My thesis of there being confusion between matching devices to buses was getting harder to justify - everything still looked sane.
+
+
+# Debuggers
+
+I do not like debuggers.  I am an avid `printk()` enthusiast.  There's no real justification for this, a bunch of my problems could almost certainly be solved easier by using actual tools, but my brain seemingly enjoys the routine of printing and building and running until I figure out what's going on.  It was becoming increasingly obvious, however, that `printk` could not save me here, and we needed to go deeper.
+
+Very thankfully for me, even though this bug was discovered on real hardware, it reproduces easily in [QEMU](https://www.qemu.org), making iteration easy.  With [GDB attached to QEMU](https://qemu-project.gitlab.io/qemu/system/gdb.html), it's time to dive in to the guts of this issue and figure out what's happening.
+
+Somehow, VIO buses are ending up with `pci_notify()` in their `bus_notifier` list.  Let's break down the data structures here with a look at `struct notifier_block`:
+
+```c
+struct notifier_block {
+        notifier_fn_t notifier_call;
+        struct notifier_block __rcu *next;
+        int priority;
+};
+```
+
+So notifier chains are [singly linked lists](https://en.wikipedia.org/wiki/Linked_list#Singly_linked_list).  When callbacks are registered through functions like `bus_register_notifier()`, after a long chain of breadcrumbs we reach [notifier_chain_register()](https://elixir.bootlin.com/linux/latest/source/kernel/notifier.c#L22) which walks the list of `->next` pointers until it reaches `NULL`, at which point it sets `->next` of the tail node to the `struct notifier_block` that was passed in.  It's very important to note here that the data being appended to the list here is *not just the callback function* (i.e. `pci_notify()`), but the `struct notifier_block` itself (i.e. `struct notifier_block pci_notifier` from earlier).  There's no new data being initialised, just updating a pointer to the object that was passed by the caller.
```

Comment:
~When~ Callbacks are registered through functions like ..., _then_ after a long chain of breadcrumbs we reach ...